### PR TITLE
Updating /driverrider endpoint and creating endpoint to register and get more info from facebook

### PR DESCRIPTION
**Pull request by olafurorng**
### What is in this PR:
- Updating /driverrider endpoint
- creating endpoint to register
- get more info from facebook (e.g. profile url and first name). I have also update the database to store this data, you can take a look at that also

@hlynurf @pajdak3

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -16,6 +16,13 @@
 
     <dependencies>
 
+        <!-- basic json objects -->
+        <dependency>
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20131018</version>
+        </dependency>
```
Comment (olafurorng) [inline pom.xml]: using this package to handle json object in java code

File: src/main/java/project/controller/DriverRiderController.java [hunk 2]
```diff
@@ -41,4 +42,37 @@ public DriverRiderController(UserService userModel, MockDataService mockDataServ
         return new ResponseEntity<DriverRiderResponse>(driverRiderResponse, HttpStatus.OK);
     }
 
+    @RequestMapping(value = "/registerrider", method = RequestMethod.POST)
```
Comment (olafurorng) [inline src/main/java/project/controller/DriverRiderController.java]: new endpoint, it should be called when creating new rider

File: src/main/java/project/controller/DriverRiderController.java [hunk 3]
```diff
@@ -41,4 +42,37 @@ public DriverRiderController(UserService userModel, MockDataService mockDataServ
...
+    public ResponseEntity<Object> registerRriver(
+            @RequestParam(value="phone_number") String phoneNumber,
+            @RequestParam(value="low_price") int lowPrice,
+            @RequestParam(value="high_price") int highPrice,
+            @RequestParam(value="number_of_people") int numberOfPeople,
+            @RequestParam(value="location") String location,
+            @RequestParam(value="destination") String destination,
+            @RequestParam(value="message") String message,
+            @RequestParam(value="pickup_time_timestamp") String pickUpTimeTimestamp
```
Comment (olafurorng) [inline src/main/java/project/controller/DriverRiderController.java]: when calling the endpoint to create new rider, you have to pass all of this data above, lines 47-54,
and the keys for the values is the variable `value`, so e.g. when sending the phone number when calling the endpoint, the key has to be `phone_number `

File: src/main/java/project/controller/DriverRiderController.java [hunk 4]
```diff
@@ -41,4 +42,37 @@ public DriverRiderController(UserService userModel, MockDataService mockDataServ
...
+    @RequestMapping(value = "/registerrider", method = RequestMethod.POST)
+    public ResponseEntity<Object> registerRriver(
+            @RequestParam(value="phone_number") int phoneNumber,
+            @RequestParam(value="low_price") int lowPrice,
+            @RequestParam(value="high_price") int highPrice,
+            @RequestParam(value="number_of_people") int numberOfPeople,
+            @RequestParam(value="location") String location,
+            @RequestParam(value="destination") String destination,
+            @RequestParam(value="message") String message,
+            @RequestParam(value="pickup_time_timestamp") long pickUpTimeTimestamp
```
Comment (olafurorng) [inline src/main/java/project/controller/DriverRiderController.java]: in the client/web you have to change the time that the user selected to timestamp,
if you don't know what timestamp is -> https://www.google.is/search?q=timestmap&oq=timestmap&aqs=chrome..69i57j69i60l3j69i59j69i61.1089j0j7&sourceid=chrome&es_sm=91&ie=UTF-8#q=what+is+timestamp

and you have to google how to change the object that you get from the user to timestamp before calling the endpoint

File: src/main/java/project/controller/DriverRiderController.java [hunk 5]
```diff
@@ -41,4 +42,37 @@ public DriverRiderController(UserService userModel, MockDataService mockDataServ
...
+            //@RequestParam(value="accessToken") String accessToken, TODO: check the accessToken
+    )
+    {
+        return new ResponseEntity<Object>(HttpStatus.CREATED);
+    }
+
+    @RequestMapping(value = "/registerdriver", method = RequestMethod.POST)
```
Comment (olafurorng) [inline src/main/java/project/controller/DriverRiderController.java]: new endpoint, it should be called when creating new driver,

File: src/main/java/project/controller/LoginController.java [hunk 6]
```diff
@@ -33,25 +33,31 @@ public LoginController(UserService userService, MockDataService mockDataService)
...
     @RequestMapping(value = "/login", method = RequestMethod.POST)
     public ResponseEntity<LoginResponse> login(
                                         @RequestParam(value="accessToken") String accessToken,
+                                        @RequestParam(value="fb_access_token") String fbAccessToken,
                                         @RequestParam(value="id") String id,
                                         @RequestParam(value="name") String name,
-                                        @RequestParam(value="email") String email
+                                        @RequestParam(value="email") String email,
+                                        @RequestParam(value="first_name") String firstName,
+                                        @RequestParam(value="picture_url") String pictureUrl
```
Comment (olafurorng) [inline src/main/java/project/controller/LoginController.java]: adding more information to database that we get from server

File: src/main/java/project/controller/LoginController.java [hunk 7]
```diff
@@ -33,25 +33,31 @@ public LoginController(UserService userService, MockDataService mockDataService)
...
+     * @param fbAccessToken access token we get from facebook
+     * @param id fb user id of the user, which we also use as user id
+     * @param name full name of user
+     * @param firstName first name of user
+     * @return
+     */
     @RequestMapping(value = "/login", method = RequestMethod.POST)
     public ResponseEntity<LoginResponse> login(
                                         @RequestParam(value="accessToken") String accessToken,
+                                        @RequestParam(value="fb_access_token") String fbAccessToken,
```
Comment (olafurorng) [inline src/main/java/project/controller/LoginController.java]: Updating the access token flow, so if the client doesn't have a server generated access token stored in the browser, the client can send the fb access token which the server will validate and if everything is correct, then the server will return the generated server access token so the client can store it in the browser.

File: src/main/java/project/controller/LoginController.java [hunk 8]
```diff
@@ -64,27 +70,19 @@ public LoginController(UserService userService, MockDataService mockDataService)
...
             if (accessTokenGeneratedFromServer == null) return new ResponseEntity<LoginResponse>(HttpStatus.SERVICE_UNAVAILABLE);
-            user = new User(accessTokenGeneratedFromServer, id, name, email, 0.0, 0, ""); // TODO: add profile picture url
+            user = new User(accessTokenGeneratedFromServer, id, name, email, 0.0, 0, pictureUrl, firstName,
+                    null); // phone number is null, as we can't get it from facebook
 
-            LoginResponse loginResponse = new LoginResponse(
-                    mockDataService.getMockDriverList(),
-                    mockDataService.getMockRiderList(),
-                    user);
+            LoginResponse loginResponse = new LoginResponse(user);
```
Comment (olafurorng) [inline src/main/java/project/controller/LoginController.java]: only returning the user for `/login` endpoint, as the client is not using the driver/rider reponse

File: src/main/java/project/mockservice/MockDataService.java [hunk 9]
```diff
@@ -35,11 +35,7 @@
             "vip9@hi.is");
 
     private final List<String> ids = Arrays.asList(
-            "7954123983290342",
-            "5434448985345409",
-            "5893763594275695",
-            "1230495844348884",
-            "5687392489905864");
+            "1177370922");
```
Comment (olafurorng) [inline src/main/java/project/mockservice/MockDataService.java]: for the mock riders and drivers for the `/driverrider` endpoint, I am now returning a valid facebook id (my fb id), so we can try out the link on the user name and profile pic, except the url is not working, a bug on the client, see task: https://app.asana.com/0/52964612936482/57748631426840

File: src/main/java/project/service/AbstractService.java [hunk 10]
```diff
@@ -37,6 +37,7 @@ protected Connection getConnectionToDatabase() throws URISyntaxException, SQLExc
         // TODO: viljum við nota '!connection.isValid(VALID_TIMEOUT)'
         if  (connection == null || connection.isClosed() || !connection.isValid(VALID_TIMEOUT))
         {
+            // TODO: use system env. variables to store database info, and when that is done, change the database user and passw @ Heroku
```
Comment (olafurorng) [inline src/main/java/project/service/AbstractService.java]: can you do this tasks when you can:
- https://app.asana.com/0/52964612936482/57722993923201
- https://app.asana.com/0/52964612936482/57722993923199
?

File: src/main/java/project/service/UserService.java [hunk 11]
```diff
@@ -84,15 +88,49 @@ public User getUser(String userId, String accessTokenFromClient) throws Unauthor
         return user;
     }
 
+
     private void checkIfExistingUserIsAuthorized(final String userId, final String accessTokenFromDatabase,
```
Comment (olafurorng) [inline src/main/java/project/service/UserService.java]: everything here below in this method is regarding -> https://github.com/pajdak3/hermes-server/pull/23/files#r41947837

we will start using the authentication like this, but I am very open to rethink it and design a new, better authentication flow

File: src/main/java/project/service/UserService.java [hunk 12]
```diff
@@ -84,15 +88,49 @@ public User getUser(String userId, String accessTokenFromClient) throws Unauthor
...
+            if (isAccessTokenFromDatabaseInvalid)
+            {
+                // lets check if fb access token is valid
+                if (fbAccessToken == null || fbAccessToken.isEmpty())
+                {
+                    throw new UnauthorizedException("an incorrect access token was sent from the web and the fb access " +
+                            "token was not sent");
+                }
+
+                /* TODO: create call to fb and check if 'fbAccessToken' is valid for userId
```
Comment (olafurorng) [inline src/main/java/project/service/UserService.java]: this is TODO,
so you can now login to other user account with fake facebook access token, I am not going to do this until we have decided/changed to new auth flow, see: https://github.com/pajdak3/hermes-server/pull/23/files#r41948046

File: src/main/webapp/WEB-INF/jsp/Main.jsp [hunk 13]
```diff
@@ -10,13 +10,13 @@
   <title>ÍsFar</title>
   <link href='https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/css/bootstrap.min.css' rel='stylesheet' type='text/css'>
   <link href='/resources/theme1/css/main.css' rel='stylesheet' type='text/css'>
-  <script src="/resources/theme1/build/jquery-1.11.3.min.js" /></script>
+  <script src="/resources/theme1/build/jquery-1.11.3.min.js"></script>
```
Comment (olafurorng) [inline src/main/webapp/WEB-INF/jsp/Main.jsp]: wrong syntax,

but I am curious about this, what stuff would you put between the  script tags? 
e.g. `<script src="/resources/theme1/build/jquery-1.11.3.min.js" >WHAT WOULD YOU PUT HERE??</script>`

File: src/main/webapp/WEB-INF/jsp/Main.jsp [hunk 14]
```diff
@@ -25,14 +25,14 @@
...
+      <h1>Mig langar að ...</h1>
     </section>
     <section class="add" id="rider">
         <a></a>
     </section>
     <section class="pickContainer">
-      <h5 class="selectRider">Ride</h5>
-      <h5 class="selectDriver">Drive</h5>
+      <h5 class="selectRider">Bjóða ísFar</h5>
+      <h5 class="selectDriver">Fá ísFar</h5>
```
Comment (olafurorng) [inline src/main/webapp/WEB-INF/jsp/Main.jsp]: translating to Icelandic, as 95-100% of users speaks icelandic,
we will probably add later translations for english as well, but until then, I am changing this to Icelandic,
If you can find a better copy, please change it :)

File: src/main/webapp/resources/theme1/src/facebook.js [hunk 15]
```diff
@@ -76,9 +76,11 @@ window.fbAsyncInit = function() {
 }(document, 'script', 'facebook-jssdk'));
 
 //  See statusChangeCallback() for when this call is made.
-function callFbGraphApi(accessToken) {
+function callFbGraphApi(fbAccessToken) {
```
Comment (olafurorng) [inline src/main/webapp/resources/theme1/src/facebook.js]: this was very unclear what access token this was, so lets try to name the variable clearly

File: src/main/java/project/controller/DriverRiderController.java [hunk 16]
```diff
@@ -30,7 +31,7 @@ public DriverRiderController(UserService userModel, MockDataService mockDataServ
     }
 
     @RequestMapping(value = "/driverrider", method = RequestMethod.GET)
```
Comment (olafurorng) [inline src/main/java/project/controller/DriverRiderController.java]: @pajdak3 now the `/driverrider` returns an User object with phone number, if we have a phone number in the database for that user, so you can use that to pre-fill the phone number when registering rider / driver

File: src/main/webapp/WEB-INF/jsp/Main.jsp (hunk 14 above)
Comment (olafurorng) [inline src/main/webapp/WEB-INF/jsp/Main.jsp]: When I think afterwards, this copy doesn't make 100% sense...
... can you also try to find some better copy, and if you do, just change it

File: src/main/webapp/WEB-INF/jsp/Main.jsp (hunk 13 above)
Comment (pajdak3) [inline src/main/webapp/WEB-INF/jsp/Main.jsp]: Myndir ekki setja neitt, erum að linka jquery javascript file-ið við okkar, þannig séð bara library. Annars þá má setja javascript kóða á milli script tagga, getur haft allan javascript kóðann í html-inu þannig, nema við erum með sér javascript skjal fyrir okkar js kóða sem við linkum síðan á.

**Pull request merged**